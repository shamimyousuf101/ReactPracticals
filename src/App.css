*{
  box-sizing: border-box;
}

html,body, #root, .App { 
  background-color:#efefef;
  height: 100%;
  margin: 0;
  padding:0;
}

  