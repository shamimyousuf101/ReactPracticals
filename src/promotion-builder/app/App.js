import React, { Component } from 'react';
import isEqual from 'lodash.isequal'

import Banner from '../common/banner/Banner';
import Content from '../common/content/Content';
import Navigation from '../common/navigation/Navigation';

import { navigationData } from '../common/navigation/navigationData';

import { promotionData } from '../../domain/promotionData';
import { getTruthyList , arrayToMap } from '../../utils/utils'

import './app.css';

let uuid4 = require('uuid4');

class App extends Component {

  state = {
    view: "Search",
    selectedPromotionId: null,
    formData: {
      id: 0,
      devices: new Map(),
      ventures: new Map(),
      url: "",
      name: "",
      lastUpdatedTime: ""
    },
    searchTerm: "",
    searchDisplay: [],
    promotionData: promotionData
  };

  

  componentDidMount(){
    this.searchPromotions();
  }


  componentDidUpdate(prevProps, prevState){
    if(!isEqual(prevState.promotionData,this.state.promotionData)){
      this.searchPromotions();
    }
    
  }

  handleInputChange = event => this.setState({ searchTerm: event.target.value });


  searchBtnClick = event => {
    event.preventDefault();
    this.searchPromotions();
  }


  searchPromotions = () => {
    const searchInput = this.state.searchTerm;
    const promotionArray = Object.keys(this.state.promotionData).map(key => this.state.promotionData[key]);
    const filteredPromotionArray = promotionArray.filter((el) => el.name.toLowerCase().indexOf(searchInput.toLowerCase()) > -1);
    this.setState({searchDisplay: filteredPromotionArray.sort((a,b) => b.lastUpdatedTime - a.lastUpdatedTime) });
  }


  onFormChange = (value, formField) => {
      this.setState(prevState => {return {  formData: { ...prevState.formData, [formField]: value }}})
  }


  savePromotion = () => {
    if (this.state.formData) {
        let id = this.getPromotionId();
        this.showAlertWithFormData();
        this.setState(prevState => {return {  promotionData: { ...prevState.promotionData, [id ]: this.getNewOrUpdatedPromotion(id)}}})
        this.setState({view: "Search"})
    }
  }


  getNewOrUpdatedPromotion = (id) => {
    const {
      formData: {
        name = '',
        url = '',
        devices = new Map(),
        ventures = new Map()
      }
    } = this.state;

    return {
      id,
      name,
      url,
      devices: [...devices.keys()],
      ventures: [...ventures.keys()],
      lastUpdatedTime: Date.now()
    };
  };


  getPromotionId = () => {
    const {
      selectedPromotionId,
      formData: {
        id: formDataId
      }
    } = this.state;

    return formDataId || selectedPromotionId || uuid4();
  }


  showAlertWithFormData = () => {
    let savedData = `New Promotion Details:
                    Name:\t ${this.state.formData.name} 
                    Url:\t ${this.state.formData.url}
                    Devices:\t ${getTruthyList(this.state.formData.devices )}
                    Ventures:\t ${getTruthyList(this.state.formData.ventures)}`;
    alert(savedData);
  }


  resetFormData = () => {
    this.setState({
      selectedPromotionId: null,
      formData: {
        devices: new Map(),
        ventures: new Map(),
        url: "",
        name: ""
      },
    });
  }


  editBtnClick = (id) => {
    this.setState({
      selectedPromotionId: id,
      view: "SavePromotion",
      formData: this.setFormDataToSelectedPromotion(id)
    });
  }


  setFormDataToSelectedPromotion = (selectedPromotionId) => {
    const selectedPromotion = this.state.searchDisplay.find(item => item.id === selectedPromotionId);
    let selectedPromotionFormData = {};
    selectedPromotionFormData.devices = arrayToMap(selectedPromotion.devices)
    selectedPromotionFormData.ventures = arrayToMap(selectedPromotion.ventures)
    selectedPromotionFormData.url = selectedPromotion.url
    selectedPromotionFormData.id = selectedPromotion.id
    selectedPromotionFormData.name = selectedPromotion.name
    return selectedPromotionFormData;
  }


  onMenuClick = (event) => {
    const selectedLink = event.currentTarget.className;
    let view;
    switch (selectedLink) {

      case "search__link":
        view = "Search";
        this.setState({searchTerm: ""}, () => {
          this.searchPromotions();
        })
        
        break;
      case "upload__link":
        view = "Upload";
        break;
      case "config__link":
        view = "SavePromotion";
        this.resetFormData();
        break;
      default:
        view = "Search";
    }

    this.setState({ view });

  }

  render() {

    const { view, formData, searchDisplay, searchTerm , selectedPromotionId} = this.state;

    return (
      <div className="App">
        <Banner subHeading={view} />
        <Content 
        view={view} 
        formData={formData} 
        selectedPromotionId={selectedPromotionId} 
        editBtnClick={this.editBtnClick} 
        reset={this.resetFormData} 
        savePromotion={this.savePromotion} 
        onFormChange={this.onFormChange} 
        handleInputChange={this.handleInputChange} 
        searchBtnClick={this.searchBtnClick} 
        searchDisplay={searchDisplay} 
        searchTerm={searchTerm} />
        <Navigation menuData={navigationData} clickHandler={this.onMenuClick} />
      </div>
    );
  }
}

export default App;

