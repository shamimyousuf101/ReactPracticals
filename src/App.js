import React, { Component } from 'react';

import Banner from './Components/Header/Banner';
import Container from './Components/DynamicContent/Container';
import Menu from './Components/Menu/Menu';
import { menuData } from './Components/Menu/menuData';

import './Components/Header/Banner.css';
import './Components/Menu/Menu.css';
import './Components/DynamicContent/Search/Search.css';
import './Components/DynamicContent/Asset/AssetManager.css';
import './Components/DynamicContent/PromotionBuilder/PromotionBuilder.css';
import './Components/DynamicContent/Container.css';
import './App.css';
import { promotionData } from './Components/DynamicContent/PromotionBuilder/promotionData';
import getTruthyList from './utils/utils'


class App extends Component {

  state = {
    view: "Search",
    selectedPromotionId: null,
    formData: {
      id: 0,
      devices: new Map(),
      ventures: new Map(),
      url: "",
      name: ""
    },
    searchTerm: "",
    searchDisplay: promotionData,
  };


  handleInputChange = event => this.setState({ searchTerm: event.target.value });

  searchBtnClick = event => {
    const searchInput = this.state.searchTerm;
    let promotionArray = Object.keys(promotionData).map(key => promotionData[key]);
    event.preventDefault();
    this.setState({
      searchDisplay: promotionArray.filter((el) => el.name.toLowerCase().indexOf(searchInput.toLowerCase()) > -1)
    });
  }


  onFormChange = (value, formField) => {
      this.setState((prevState) => {
        return {  formData: { ...prevState.formData, [formField]: value }}
      })
  }



  savePromotion = () => {


    // BUG - new id getting created each time save is clicked

    function MapToArray(map) {
      
      let arr1 = [...map.keys()]
      return arr1;
    }

    let uuid4 = require('uuid4');
    let id = null;
    let newPromotion = {
      id: 0,
      devices: [],
      ventures: [],
      url: "",
      name: ""
    }

    if (this.state.formData) {

        

        if(this.state.selectedPromotionId){
          id = this.state.selectedPromotionId
        }

        if(this.state.formData.id){
            id = this.state.formData.id
        } else {
          id = uuid4();
        }

        let data = " New Promotion Details: " + "\n" + 
        "Name: " + this.state.formData.name + "\n" + 
        "Id: " + id + "\n" + 
        "Url: " + this.state.formData.url + "\n" + 
        "Devices: " + getTruthyList(this.state.formData.devices ) + "\n" + 
        "Ventures: " + getTruthyList(this.state.formData.ventures) ;
        alert(data);

        newPromotion.id=id;
        newPromotion.name=this.state.formData.name;
        newPromotion.url=this.state.formData.url;
        newPromotion.devices=MapToArray(this.state.formData.devices);
        newPromotion.ventures=MapToArray(this.state.formData.ventures);

        this.setState((prevState) => {
          return {  formData: { ...prevState.formData, [id]: newPromotion }}
        })

      this.setState((prevState) => {
        return {  searchDisplay: { 
          ...prevState.searchDisplay, 
          [id ]: newPromotion
        }}
      })

    }

}

  reset = () => {
    this.setState({
      selectedPromotionId: null,
      formData: {
        devices: new Map(),
        ventures: new Map(),
        url: "",
        name: ""
      },
    });
  }


  editBtnClick = (id) => {
    this.setState({
      selectedPromotionId: id,
      view: "PromotionBuilder",
      formData: this.setFormDataToSelectedPromotion(id)
    });
  }


  setFormDataToSelectedPromotion = (selectedPromotionId) => {

    let selectedPromotion = this.state.searchDisplay.find(item => item.id === selectedPromotionId);
    let selectedPromotionFormData = {};
    selectedPromotionFormData.devices = arrayToMap(selectedPromotion.devices)
    selectedPromotionFormData.ventures = arrayToMap(selectedPromotion.ventures)
    selectedPromotionFormData.url = selectedPromotion.url
    selectedPromotionFormData.id = selectedPromotion.id
    selectedPromotionFormData.name = selectedPromotion.name
    return selectedPromotionFormData;

    function arrayToMap(array) {
      let newMap = new Map();

      if (array.length > 0) {
        array.forEach(element => {
          newMap.set(element, true);
        });
      }
      return newMap;
    }

  }




  onclick = (event) => {
    const selectedLink = event.currentTarget.className;
    let view;
    switch (selectedLink) {

      case "search__link":
        view = "Search";
        break;
      case "upload__link":
        view = "Upload";
        break;
      case "config__link":
        view = "PromotionBuilder";
        this.reset();
        break;
      default:
        view = "Search";

    }

    this.setState({ view });

  }

  render() {

    const { view, formData, searchDisplay, searchTerm } = this.state;

    return (
      <div className="App">
        <Banner subHeading={view} />
        <Container view={view} formData={formData} editBtnClick={this.editBtnClick} reset={this.reset} savePromotion={this.savePromotion} onFormChange={this.onFormChange} handleInputChange={this.handleInputChange} searchBtnClick={this.searchBtnClick} searchDisplay={searchDisplay} searchTerm={searchTerm} />
        <Menu menuData={menuData} clickHandler={this.onclick} />
      </div>
    );
  }
}

export default App;

