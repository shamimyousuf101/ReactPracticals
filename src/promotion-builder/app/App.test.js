import React from "react";
import { shallow } from "enzyme";
import renderer from 'react-test-renderer';

import App from "./App";


describe("App component", () => { 
  it('renders the html correctly', () => {
    const tree = renderer
      .create(<App/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the necessary Components only once", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("Navigation")).toHaveLength(1);
    expect(wrapper.find("Banner")).toHaveLength(1);
    expect(wrapper.find("Search")).toHaveLength(1);
    expect(wrapper.find("AssetManager")).toHaveLength(1);
    expect(wrapper.find("SavePromotion")).toHaveLength(1);
  });

});
