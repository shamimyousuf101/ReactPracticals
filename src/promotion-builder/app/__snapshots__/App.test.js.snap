// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
<div
  className="promotion"
>
  <header
    className="banner"
  >
    <h1>
      Promotion Manager
    </h1>
    <h3>
      Search
    </h3>
  </header>
  <nav
    className="navigation"
  >
    <ul>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Search
        </a>
      </li>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Asset Manager
        </a>
      </li>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Create / Edit
        </a>
      </li>
    </ul>
  </nav>
  <main
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <section
      className="searchContainer"
    >
      <form
        className="search__form"
      >
        <img
          alt="search"
          className="button__search"
          src="/search.png"
        />
        <input
          className="search__input"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </form>
      <ul>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              welcome back
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                welcome back url  January 2, 2019 9:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile,tablet,desktop
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                jackpotjoy,botemania
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              welcome
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                welcome url January 2, 2019 7:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile,tablet
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                jackpotjoy,starspins
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              lapse
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                lapse url January 2, 2019 6:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                starspins,heartbingo
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
      </ul>
    </section>
  </main>
</div>
`;

exports[`App component renders the html correctly 1`] = `
<div
  className="promotion"
>
  <header
    className="banner"
  >
    <h1>
      Promotion Manager
    </h1>
    <h3>
      Search
    </h3>
  </header>
  <nav
    className="navigation"
  >
    <ul>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Search
        </a>
      </li>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Asset Manager
        </a>
      </li>
      <li
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Create / Edit
        </a>
      </li>
    </ul>
  </nav>
  <main
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <section
      className="searchContainer"
    >
      <form
        className="search__form"
      >
        <img
          alt="search"
          className="button__search"
          src="/search.png"
        />
        <input
          className="search__input"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </form>
      <ul>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              welcome back
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                welcome back url  January 2, 2019 9:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile,tablet,desktop
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                jackpotjoy,botemania
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              welcome
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                welcome url January 2, 2019 7:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile,tablet
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                jackpotjoy,starspins
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
        <li
          className="singlesearchresult"
        >
          <details
            className="promotionDetails"
          >
            <summary>
              lapse
            </summary>
            <dl>
              <dt>
                Url:
              </dt>
              <dd>
                lapse url January 2, 2019 6:31:53 AM
              </dd>
              <dt>
                Devices:
              </dt>
              <dd>
                mobile
              </dd>
              <dt>
                Ventures:
              </dt>
              <dd>
                starspins,heartbingo
              </dd>
            </dl>
          </details>
          <button
            className="edit__button"
            onClick={[Function]}
            type="submit"
          >
            Edit
          </button>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
